[GSoC] API change for the newly added system of ODEs solvers

mijo2 opened the pull request:
#### References to other Issues or PRs
API change for the system of ODEs solvers introduced in:
PR #18720 - Type 1
PR #19341 - Type 2
PR #19185 - Type 3
PR #19594 - Type 4

#### Brief description of what is fixed or changed
The API for the 4 new linear first-order system of ODEs solvers was changed and these solvers were combined into one in this PR and thus, it can now be accessed directly by anyone without using the matching dictionary as was used previously.

#### Other comments
An end-user can directly call `linodesolve` function to get the solution if the user has the coefficient matrix and the non-homogeneous term(if any) available.

#### Release Notes

<!-- BEGIN RELEASE NOTES -->
* solvers
  * API for the _linear_neq_order1_type1-4 was changed to make it easier for anyone to access the solver
<!-- END RELEASE NOTES -->

mijo2 committed:
solvers.ode: Changed the API for the newly added solvers

Changed the API, the function that invokes the solver was also
changed so that the dsolve method can handle the cases properly
and a minor test case change was done in this commit

mijo2 committed:
solvers.ode: Added test cases for checking the error handling

Error handling test cases were added in this commit for all the
newly added solvers

oscarbenjamin commented:
The solver functions for the 4 cases are very similar. Is it possible to combine them into a single function?

The rhs argument could be made optional (defaulting to a zero vector). There could be an optional argument to specify whether it is type1 etc. If the type is not provided then the matcher function could be called to determine the type.

The docstring could demonstrate how to use the function in combination with `linear_ode_to_matrix`.

mijo2 commented:
I was thinking of something similar but there is one thing that needs to be taken care of, if we follow exactly the steps that you have outlined, then how would we handle the case where the non constant coefficient matrix is not commutative with its antiderivative?

For now, we are handling those things by just using the matching function. It appropriately returns `None` and thus it becomes easier for us to pass the system onto the older solvers. But if we follow your proposed strategy, then we need to add some kind of well thought structure so that the systems which can't be handled are smoothly passed on as it is getting passed on now.

oscarbenjamin commented:
The matching function can still return `None` and can be used in `classifysysode`. The solver function can just raise `NotImplementedError` in that situation. It can have a docstring that explains how to use it with `linear_ode_to_matrix` and explains the 4 types of system that are supported. Additionally if the type is supplied by the caller then there doesn't need to be any checking to see if the type is correct - the docstring should clarify that the results will not necessarily be correct if the type specified is incorrect.

mijo2 commented:
Thanks for clearing it, I would start working on it soon.

mijo2 committed:
solvers.ode: Added the cumulative solver

This solver sums up all the solvers together and this can be easily
used by an end-user as well since the API is designed that way.
This function was designed because there were a lot of repetitive
lines in the solvers.

mijo2 commented:
I have designed the first version of the general solver, I would check if there are any flaws with it for now and later incorporate it into our code base, add test cases and documentation.

mijo2 committed:
solvers.ode: Added docstring for _linear_neq_order1_solver

mijo2 committed:
solvers.ode: Incorporated the new combined solver in dsolve

mijo2 committed:
solvers.ode: Added test cases for the new solver

The test cases are mainly to check if the function raises the
appropriate errors. Solution checking is done with the use of
dsolve.

mijo2 committed:
doc: Removed the old solver documentation and added the new one

mijo2 committed:
solvers.ode: Added documentation for _canonical_equations

mijo2 committed:
solvers.ode: Added documentation for the commutative check function

mijo2 committed:
solvers.ode: Removed the old solvers code

mijo2 commented:
@oscarbenjamin I think the combined solver is ready for now, I have added it and removed the four solvers introduced previously. Along with that, I have added documentation for two helper functions: `is_commutative_anti_derivative` and `_canonical_equations`.

mijo2 committed:
solvers.ode: Added test cases for _canonical_equations

Added test cases for better coverage

mijo2 committed:
solvers.ode: Added a fix for an issue and added test cases

Added test cases for doit argument for better coverage.

oscarbenjamin commented on sympy/solvers/ode/systems.py:
```diff
@@ -408,251 +409,382 @@ def jordan_chains(A):
+        P = simplify(P)
+
+        sol_vector = P * (J * Cvect) if type == "type1" else P * J * ((J.inv() * P.inv() * b).applyfunc(
+            lambda x: Integral(x, t)) + Cvect)
```
It's better not to use inline if/else unless the expressions are very short. This should be 
```
if type == "type1":
   ...
elif ...
```

oscarbenjamin commented:
I think we should make `_linear_neq_order1_solver` a public function called something like `solve_linear_ode_system` or perhaps `linodesolve` by analogy with `linsolve`.

I also think that `_canonical_odes` should be a public function. It should solve the equations algebraically for the highest-order derivative of each unknown function. In general there can be multiple solutions so it should return a list of possible results. Note that there can be multiple solutions with each being linear e.g.:
```julia
In [12]: eqs = [Eq(f(x).diff(x)**2, g(x).diff(x)**2), Eq(f(x).diff(x), f(x))]                                                                  

In [13]: eqs                                                                                                                                   
Out[13]: 
⎡          2             2                 ⎤
⎢⎛d       ⎞    ⎛d       ⎞   d              ⎥
⎢⎜──(f(x))⎟  = ⎜──(g(x))⎟ , ──(f(x)) = f(x)⎥
⎣⎝dx      ⎠    ⎝dx      ⎠   dx             ⎦

In [14]: solve(eqs, [f(x).diff(x), g(x).diff(x)], dict=True)                                                                                   
Out[14]: 
⎡⎧d               d              ⎫  ⎧d               d             ⎫⎤
⎢⎨──(f(x)): f(x), ──(g(x)): -f(x)⎬, ⎨──(f(x)): f(x), ──(g(x)): f(x)⎬⎥
⎣⎩dx              dx             ⎭  ⎩dx              dx            ⎭⎦
```
Ideally `dsolve` should be able to handle that case and return multiple solutions. I think `_canonical_equations` should be happy to return results even in the nonlinear case if solve is able to solve the equations algebraically. It should be up to the caller to decide if the equations are linear or not.

Making these public means giving them names that don't start with underscore and also adding them in the `__init__.py` at least for the ode module. Together this gives a set of functions that can be used by users if they want:
```
canonical_odes
linear_ode_to_matrix
linodesolve
```
Alternatively they can use `dsolve` which combines those together.

mijo2 committed:
solvers.ode: Some changes in docstring and code

Long inline if/else were dropped to normal if else conditions and
_canonical_equations and _is_commutative_anti_derivative references
are added in new solver's docstring.

oscarbenjamin commented:
I think it would be better to call `checksysodesol` in `checkodesol` and document that users can use `checkodesol`.

mijo2 commented:
> Making these public means giving them names that don't start with underscore and also adding them in the `__init__.py` at least for the ode module.

Ok, makes sense, I would start working on it right away.

> Note that there can be multiple solutions with each being linear

In that case, we can maybe return a list of lists where each list is a solution for the system of ODEs passed. We would have to integrate it well later on when we add weakly and strongly connected components and higher-order to lower-order reduction. But yes, this will give us a chance to work with simplistic non-linear system of ODEs.

oscarbenjamin commented:
This is nice:
```julia
In [53]: from sympy.solvers.ode.systems import _canonical_equations as canonical_ode_system, linear_ode_to_matrix, _linear_neq_order1_solver as
    ...:  linodesolve

In [44]: func = [f(x), g(x)]                                                                                                                   

In [45]: eq = [Eq(f(x).diff(x)+g(x).diff(x), g(x)), Eq(g(x).diff(x), f(x))]                                                                    

In [46]: eq                                                                                                                                    
Out[46]: 
⎡d          d                d              ⎤
⎢──(f(x)) + ──(g(x)) = g(x), ──(g(x)) = f(x)⎥
⎣dx         dx               dx             ⎦

In [47]: eq2 = canonical_ode_system(eq, func, x)                                                                                               

In [48]: eq2                                                                                                                                   
Out[48]: 
⎡d                        d              ⎤
⎢──(f(x)) = -f(x) + g(x), ──(g(x)) = f(x)⎥
⎣dx                       dx             ⎦

In [49]: (A, B), rhs = linear_ode_to_matrix(eq2, func, x, 1)                                                                                   

In [50]: linodesolve(B, x, type="type1")                                                                                                       
Out[50]: 
⎡                 ⎛1   √5⎞                    ⎛1   √5⎞        ⎛1   √5⎞         ⎛1   √5⎞⎤
⎢               x⋅⎜─ - ──⎟                  x⋅⎜─ + ──⎟      x⋅⎜─ - ──⎟       x⋅⎜─ + ──⎟⎥
⎢   ⎛  1   √5⎞    ⎝2   2 ⎠      ⎛  √5   1⎞    ⎝2   2 ⎠        ⎝2   2 ⎠         ⎝2   2 ⎠⎥
⎢C₁⋅⎜- ─ + ──⎟⋅ℯ           + C₂⋅⎜- ── - ─⎟⋅ℯ          , C₁⋅ℯ           + C₂⋅ℯ          ⎥
⎣   ⎝  2   2 ⎠                  ⎝  2    2⎠                                             ⎦
```
If I don't specify type1 though it gives a more complicated result:
```julia
In [52]: linodesolve(B, x)                                                                                                                     
Out[52]: 
⎡                                                                                                                 √5⋅x   x                x  
⎢           x   √5⋅x              √5⋅x   x               √5⋅x   x           √5⋅x   x          x   √5⋅x          - ──── + ─                ─ +
⎢           ─ + ────            - ──── + ─             - ──── + ─         - ──── + ─          ─ + ────       2     2     2 ⌠              2  
⎢           2    2                 2     2          2     2     2            2     2          2    2     √5⋅x ⋅ℯ          ⋅⎮ 0 dx   √5⋅x⋅ℯ   
⎢  √5⋅C₁⋅x⋅ℯ           √5⋅C₁⋅x⋅ℯ             √5⋅C₂⋅x ⋅ℯ             C₂⋅x⋅ℯ             √5⋅C₂⋅ℯ                             ⌡                 
⎢- ───────────────── + ─────────────────── - ──────────────────── + ──────────────── - ─────────────── - ──────────────────────── - ─────────
⎢      ⎛  √5⋅x   x⎞          ⎛x   √5⋅x⎞                     2           x   √5⋅x              5                           2               ⎛  
⎢    5⋅⎜- ──── + ─⎟        5⋅⎜─ + ────⎟           ⎛x   √5⋅x⎞            ─ + ────                                ⎛x   √5⋅x⎞              5⋅⎜- 
⎢      ⎝   2     2⎠          ⎝2    2  ⎠         5⋅⎜─ + ────⎟            2    2                                5⋅⎜─ + ────⎟                ⎝  
⎣                                                 ⎝2    2  ⎠                                                    ⎝2    2  ⎠                   

 √5⋅x                  √5⋅x   x               √5⋅x   x              x   √5⋅x                                                                 
 ────                - ──── + ─             - ──── + ─              ─ + ────                  x   √5⋅x            √5⋅x   x            x   √5⋅
  2   ⌠                 2     2 ⌠              2     2 ⌠            2    2   ⌠                ─ + ────          - ──── + ─            ─ + ───
     ⋅⎮ 0 dx   √5⋅x⋅ℯ          ⋅⎮ 0 dx   x⋅ℯ          ⋅⎮ 0 dx   √5⋅ℯ        ⋅⎮ 0 dx           2    2               2     2            2    2 
      ⌡                         ⌡                      ⌡                     ⌡         √5⋅C₁⋅ℯ           √5⋅C₁⋅ℯ             √5⋅C₂⋅x⋅ℯ       
──────────── + ─────────────────────── + ──────────────────── - ───────────────────, - ─────────────── + ───────────────── + ────────────────
√5⋅x   x⎞              ⎛x   √5⋅x⎞              x   √5⋅x                  5                    5                  5                ⎛x   √5⋅x⎞ 
──── + ─⎟            5⋅⎜─ + ────⎟              ─ + ────                                                                         5⋅⎜─ + ────⎟ 
 2     2⎠              ⎝2    2  ⎠              2    2                                                                             ⎝2    2  ⎠ 
                                                                                                                                             

                                                 x   √5⋅x                  √5⋅x   x              x   √5⋅x                √5⋅x   x            
x              √5⋅x   x                          ─ + ────                - ──── + ─              ─ + ────              - ──── + ─            
─            - ──── + ─         √5⋅x   x         2    2   ⌠                 2     2 ⌠            2    2   ⌠               2     2 ⌠          
                2     2       - ──── + ─   √5⋅x⋅ℯ        ⋅⎮ 0 dx   √5⋅x⋅ℯ          ⋅⎮ 0 dx   √5⋅ℯ        ⋅⎮ 0 dx   √5⋅ℯ          ⋅⎮ 0 dx    -
    √5⋅C₂⋅x⋅ℯ                    2     2                  ⌡                         ⌡                     ⌡                       ⌡          
─ - ─────────────────── + C₂⋅ℯ           + ───────────────────── - ─────────────────────── - ─────────────────── + ───────────────────── + ℯ 
          ⎛x   √5⋅x⎞                              ⎛x   √5⋅x⎞               ⎛x   √5⋅x⎞                 5                      5               
        5⋅⎜─ + ────⎟                            5⋅⎜─ + ────⎟             5⋅⎜─ + ────⎟                                                        
          ⎝2    2  ⎠                              ⎝2    2  ⎠               ⎝2    2  ⎠                                                        
                                                                                                                                             

                ⎤
                ⎥
 √5⋅x   x       ⎥
 ──── + ─       ⎥
  2     2 ⌠     ⎥
         ⋅⎮ 0 dx⎥
          ⌡     ⎥
                ⎥
                ⎥
                ⎦
```

mijo2 commented:
> If I don't specify type1 though it gives a more complicated result:

For now, if the user doesn't specify the type of the system, then the solver will assume `type4` since its the most general form and can solve the systems of all the four types. But, I think it's easy to add an `auto` option, wherein the solver will determine the type of the system. I will add that option soon.

mijo2 commented:
@oscarbenjamin Also, one more thing. If the user passes a `type3` or `type4` system of ODEs, and if we evaluate the antiderivative of the coefficient matrix and it is found that the system doesn't have commutative antiderivative, then should we raise a `NotImplementedError`? 

If the user passes the antiderivative, regardless of whether it is commutative with the coefficient matrix or not, the solver must give the solution, correct or incorrect.

mijo2 commented:
> `linodesolve(B, x, type="type1")`

The user would have to pass `-B`.

mijo2 committed:
solvers.ode: Added auto option in the new solver

mijo2 committed:
solvers.ode: Renamed _linear_neq_order1_solver to linodesolve

oscarbenjamin commented:
I think that the if type is not specified the the default should be to call the function that classifies this as e.g. type1. That should also be made a public function I think.

Otherwise if the type is specified (e.g. when this is called as part of dsolve) then we should assume that the classifying function was already called and not spend any time checking anything unless it's a very quick check.

mijo2 commented:
> I think that the if type is not specified the the default should be to call the function that classifies this as e.g. type1. That should also be made a public function I think.

For now, we were classifying the system of ODEs using the logic defined in the matching function. It makes sense to create a function that classifies the type by taking in two boolean values. I would create this type of function soon.

mijo2 committed:
solvers.ode, doc: Updated the private function to a public one

_canonical_equations was renamed and updated to canonical_odes.
Along with that, it is now made to handle Nonlinear and higher
order ODEs as well.

mijo2 committed:
solvers.ode: Added a return_system_type function

This function takes in the details for a system and then returns
the appropriate type.

oscarbenjamin commented:
I think that for users we can have a function that takes the two matrices and reports whether the result is type1, type2, ...

mijo2 committed:
solvers.ode: Added return_system_type function for info

System info is returned by this function based on the coefficient
matrix, independent variable and the non-homogeneous term if passed

mijo2 committed:
solvers.ode: Added three functions from systems in __init__.py

canonical_odes, linear_ode_to_matrix, linodesolve were added in
solvers/ode/__init__.py and changed the imports in systems.py

mijo2 commented:
@oscarbenjamin I have almost made the changes you asked for. Now our solver can handle system of ODEs, which can be non-linear but may be represented as multiple linear first order system of ODEs. Except the `checksysode` part, rest of the suggestions are nearly implemented. You can have a look.

oscarbenjamin commented:
It's looking good so far. What else do you want to do before this is merged?

This is already quite big so the work could be continued in another PR.

mijo2 commented:
Just a few things like:

1. Maybe rename the `return_systems_type` to `return_systems_info` since it makes more sense that way.
2. Add doctring examples in `linodesolve` that shows how a user can use the function in conjunction with `return_system_info`.
3. Should I add `checksysodesol` to `checkodesol` and update the docstring in `linodesolve` to use `checkodesol` and remove `checksysodesol` documentation added in `ode.rst`

oscarbenjamin commented:
> Just a few things like:
> 
> 1. Maybe rename the `return_systems_type` to `return_systems_info` since it makes more sense that way.

I suggest `linodesolve_type` to make it clear that the types are those as defined by `linodesolve`.

> 3. Should I add `checksysodesol` to `checkodesol` and update the docstring in `linodesolve` to use `checkodesol` and remove `checksysodesol` documentation added in `ode.rst`

Yes, but maybe in a followup PR.

oscarbenjamin commented:
Also since these functions all go together it would be good to have them refer to each other in a "see also" section of the docstring.

mijo2 committed:
solvers.ode, doc: Renamed the information function

Renamed it to linodesolve_type and added it in the docs

mijo2 committed:
solvers.ode, doc: Added test cases and updated the documentation

Added more test cases for better coverage and moved the position
of checksysodesol's docstring in ode.rst

mijo2 commented:
@oscarbenjamin After the tests have passed, we can merge this PR.

mijo2 committed:
Merge branch 'master' into dsystem_solver_api

mijo2 committed:
solvers.ode: Updated _simpsol and _solsimp tests

oscarbenjamin commented on sympy/solvers/ode/ode.py:
```diff
@@ -574,6 +574,19 @@ def dsolve(eq, func=None, hint="default", simplify=True,
+        # ODEs with the help of canonical_odes function. Hence, the
+        # linodesolve can be directly used to get the solution in
+        # this case. In future, this may have to be changed.
+        if isinstance(match, list):
```
This seems a bit messy. It's okay for now but needs to be sorted out later:

1. We shouldn't add special cases in the dsolve function itself.

2. Actually all of the systems code should be moved out from here into a separate `dsolve_system` function.

3. Properly handling multiple solutions/equations in a consistent way meaning using lists everywhere. We shouldn't have `classify_sysode` sometimes return a list and sometimes not.

4. I didn't necessarily mean that we should support these in dsolve right now but just that `canonical_odes` should be able to handle the nonlinear case.

5. The fact that dsolve for single ODEs sometimes returns an Eq and sometimes returns a list of Eq is a mistake IMO. The same happens for systems in that a single solution is a list of Eq but with multiple solutions we return a list of list of Eq (also a mistake). We need to keep that for backwards compatibility but I think that we should make sure that all internal code always operates on lists of solutions. Extracting a single solution if there is only one should be done here in dsolve right before returning the solution to the user.

mijo2 replied:
> Actually all of the systems code should be moved out from here into a separate dsolve_system function.

That's the plan for now, to get rid of the special case that I have added in this PR in `dsolve` code in the next PR.

Ok so basically it would be better if we return a list of lists always even when there is one solution. That makes sense.

Should I incorporate these changes in this PR itself?

oscarbenjamin replied:
> Ok so basically it would be better if we return a list of lists always even when there is one solution. That makes sense.

We can't change dsolve to do this because of backwards compatibility. However we can make sure that internal code always does. If `dsolve` calls a new function `dsolve_system` then `dsolve_system` can always return a list of solutions even if there is only one.

Where the system is nonlinear but can be solved to a finite number of systems in explicit form by `canonical_odes` then I think `classify_sysode` should just say something like "this is an algebraically solvable implicit system". The handler for that case can recursively call `dsolve_system` with the individual explicit systems that can then be passed to `classify_sysode` to see if they are linear etc. That way `classify_sysode` always takes a single system and gives a single classification.

mijo2 replied:
Yes I agree with you, the `dsolve` code is very large to meddle even a small part of it
 I have tried some experimental changes in `dsolve `code but all led to failure of some part. And it does make sense to always output a list of lists internally to properly represent multiple solutions.

Perhaps this work can be done in the next PR, where the major target would be to finally include `dsolve_system`. Firstly, I would try to cherry pick parts of the planned structure that I mentioned in my GSoC proposal and implement that structure keeping in mind what we have discussed in these 5 PRs related to the project.

oscarbenjamin commented:
Looks good. I like that there is a set of functions that can be used in composition by users as well as being used together in `dsolve`. I think we should make sure that we do something similar e.g. with being able to convert higher-order to 1st order so that users can use that directly and use it in combination with `linear_ode_to_matrix` etc.

There are some inconsistencies though as e.g. `linear_ode_to_matrix` works with equations in the form
```
A0 X' + A1 X = b
```
whereas `canonical_odes` and `linodesolve` use the form
```
X' = A X + b
```
so there is a minus sign difference between A and A1. Also `linear_ode_to_matrix` works with works with a list of `A` matrices but `linodesolve` works with a single matrix. We might later want to add the support for higher-order systems in `linodesolve` and then we might think that it is better to have a common representation of systems used by both `linodesolve` and `linear_ode_to_matrix`.

At some point we should make a special page in the docs for systems of ODEs and move all the systems related functions there.
